File: src/0456_132_pattern.js
/**
 * @param {number[]} nums
 * @return {boolean}
 */
var find132pattern = function (nums) {
    let stack = []; // [num, minLeft]
    let curMin = nums[0];

    for (n of nums.slice(1)) {
        while (stack.length > 0 && n >= stack.at(-1)[0]) {
            stack.pop();
        }
        if (stack.length > 0 && n > stack.at(-1)[1]) {
            return true;
        }

        stack.push([n, curMin]);
        curMin = Math.min(curMin, n);
    }

    return false;
};

module.exports = { find132pattern };

if (require.main === module) {
    const assert = require('assert');
    function ListNode(val = 0, next = null) { this.val = val; this.next = next; }
    function TreeNode(val = 0, left = null, right = null) { this.val = val; this.left = left; this.right = right; }
    function Node(val = 0, neighbors = []) { this.val = val; this.neighbors = neighbors; }
    function listNodeToString(node) {
        const values = [];
        let current = node;
        while (current !== null) {
            values.push(String(current.val));
            current = current.next;
        }
        return values.join('->');
    }
    function treeToString(node) {
        if (node === null) return 'None';
        return `${node.val}(${treeToString(node.left)},${treeToString(node.right)})`;
    }
    function sorted(value) { return value.slice().sort(); }
    assert.deepStrictEqual(find132pattern([1, 2, 3, 4]), false);
    assert.deepStrictEqual(find132pattern([3, 1, 4, 2]), true);
}

File: src/0456_132_pattern.rs
struct Solution;

impl Solution {
    pub fn find132pattern(nums: Vec<i32>) -> bool {
        let mut stack: Vec<(i32, i32)> = vec![]; // (num, min_left)
        let mut current_min = nums[0];

        for n in nums.iter().skip(1) {
            while !stack.is_empty() && *n >= stack.last().unwrap().0 {
                stack.pop();
            }
            if !stack.is_empty() && *n > stack.last().unwrap().1 {
                return true;
            }

            stack.push((*n, current_min));
            current_min = current_min.min(*n);
        }

        false
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    #[test]
    fn main_asserts() {
        assert_eq!(Solution::find132pattern(vec![1, 2, 3, 4]), false);
        assert_eq!(Solution::find132pattern(vec![3, 1, 4, 2]), true);
    }
}
